### PR TITLE
New toolbar

### PR DESCRIPTION
**Pull request by pragyavaishanav**
update on new tool bar

### CODE REVIEW

**Comment by luxaritas** [comment]
Got the merge conflicts handled and tried to get the autohide working as expected again (which was the bulk of where the conflicts originated). I'll deploy to dev on Tuesday for functionality testing, at which point I'll also work on code review.

File: src/eterna/ui/Toolbar.ts [hunk 1]
```diff
@@ -145,676 +386,3015 @@ export default class Toolbar extends ContainerObject {
...
+        this.makeCustomButtons();
+        this.makeTargetButtons();
+
+        this.makeInfoLayout();
+        this.makeSolveLayout();
+        this.makeCreateLayout();
+        this.makeViewLayout();
+        this.makeInoutLayout();
+        this.makeAnnotateLayout();
+        this.makeCustomLayout();
```
Comment (luxaritas) [inline src/eterna/ui/Toolbar.ts:2512]: Why are making the buttons and the layouts two separate steps? Why not just "make<X>Category"?

File: src/eterna/ui/Toolbar.ts [hunk 2]
```diff
@@ -145,676 +386,3015 @@ export default class Toolbar extends ContainerObject {
...
-            .up(Bitmaps.ImgUndo)
-            .over(Bitmaps.ImgUndoOver)
-            .down(Bitmaps.ImgUndoHit)
-            .tooltip('Undo (Z)')
-            .hotkey(KeyCode.KeyZ)
-            .rscriptID(RScriptUIElementID.UNDO);
+            // this._scrollContainer.content.addChild(this._draggingElement)
+            // if (this._currentTab.category === this._getButtonCategory(this._draggingElement)) {
+            //     // this._scrollContainer.content.addChild(this._draggingElement);//kkk
+            // }
```
Comment (luxaritas) [inline src/eterna/ui/Toolbar.ts:3156]: Dead code, remove?

File: src/eterna/ui/Toolbar.ts [hunk 3]
```diff
@@ -145,676 +386,3015 @@ export default class Toolbar extends ContainerObject {
...
-            endScroll();
-            scrollLeft();
-        }));
+    private collapse() {
+        this._isExpanded = false;
+        this.collapseButton.container.visible = false;
+        this.expandButton.container.visible = true;
+        this._tabsHContainer.visible = false;
+
+        this.removeNamedObjects('test');
```
Comment (luxaritas) [inline src/eterna/ui/Toolbar.ts:3334]: Probably want to give it a better name - and probably use a constant like is done for the location task in the autohide handler

File: src/eterna/ui/Toolbar.ts [hunk 4]
```diff
@@ -145,676 +386,3015 @@ export default class Toolbar extends ContainerObject {
...
+                                })
+                            );
+                            this.dynPaintTools.push(button);
+                            button.display.interactive = true;
+                            this.pushButtonToCategory(button);
+                        });
+                    }
+                }
+            }
+            this.boosterButtonsGroup = new ButtonsGroup(boosterBts);
```
Comment (luxaritas) [inline src/eterna/ui/Toolbar.ts:1783]: Why create a separate group for boosters? The booster button should just be like any other tool, able to be put into either the left or right bay

**Review by luxaritas — CHANGES_REQUESTED** [review]

**Comment by sbutkovi** [comment]
Regarding the updated upper-left portion of the screen (see screenshot below):

- It looks like the colored background behind the folding engine text (and also the Natural/Target Delta box, although I don't think that was changed recently?) isn't centered correctly - it seems shorter on the right side
- Design-wise, I want to make sure that the lack of a colored background behind natural mode/target mode is desired (as it differs from how it was on the old toolbar)
- The natural mode, target mode, and switch (not sure what the official name for this one is, but it's "1/1" in this screenshot) buttons do not have tooltips that explain what they are - I envision this could be confusing particularly for newer players

pic1

**Comment by luxaritas** [comment]
Behind the natural/target mode there should be backgrounds, per the comp. I'll also add that the switch state toggle should not be present if there's only one state

**Comment by sbutkovi** [comment]
Regarding the settings menu:

- There is a "maximize" box at the upper-left of the window, but because it doesn't do anything when the window is "normal" size (besides flashing once to show someone clicked on it), it's not easily clear to players what this button does.

pic1

- The window can be resized, but upon first resizing it seems that this is the smallest the window can get:

pic2

- However, with a lot of poking and prodding (alternating between trying to stretch the various stretchable corners), it is possible to make the window smaller, but only by a few pixels each time (see below). There is similar behavior when trying to increase the size of the window - it can only be stretched larger by a few pixels at a time and works most consistently when alternating between corners.

pic3

**Comment by sbutkovi** [comment]
The settings menu does not recenter when the browser window is resized (here, the browser window was made smaller):

pic1

Compare to the behavior of the current settings menu - although it also overlaps with the toolbar, it does recenter itself:

pic2

**Comment by sbutkovi** [comment]
When the browser window is vertically small, some of the game elements overlap with each other. Notably, the natural/target mode buttons (and everything below them within the game) are not visible, and there is no way for the user to "scroll" and see them (which could potentially be an issue for small screens, such as mobile screens).

pic1

When the toolbar is expanded on this vertically small screen, the toolbar options can overlap strongly with the icons on the left-hand side, making the screen difficult to interpret. Additionally, the logic behind what is in "front" when elements overlap is not obvious - some of the left-hand-side elements appear over the toolbar options, but not all of them (such as energy values).

pic2

**Comment by sbutkovi** [comment]
- When I hover over the text that states the folding engine, my mouse cursor changes into the "do not" symbol - it just seems unusual because no other game elements seem to trigger that behavior.
- There's something jarring about the text and/or fonts when Quick Help is brought up, but I can't quite put my finger on what. Are there multiple fonts being used? (I know that the bottoms of some of the letters (mostly g) are cut off in the labels, but I don't think that's it)

pic2

- The screen gets crowded when Quick Help and the toolbar are both opened simultaneously:

pic3

- When icons are added to the toolbar, there are pop-up windows that say whether the icon will be inserted or will replace another icon. However, it does not seem possible to rearrange buttons that are already on the toolbar.
- From my understanding, the icons that are already on the toolbar are supposed to be "greyed out" in the selection menu, but it does not appear to be working completely as intended. In the following screenshot, the boosters icon (lightning bolt) is indeed greyed out, but the undo, redo, and shift buttons do not appear greyed out:

pic4

- The color behavior of the palette when the toolbar is extended is inconsistent. Note how the following screenshot has typical palette coloring, but the previous screenshot has the palette greyed out. Along those lines, it seems odd that the boosters icon is no longer greyed out in the extended menu . . . (see previous point)

pic6

- Is the boosters button intended to (1) always be on the toolbar, and (2) be in its own uneditable section of the toolbar? If so, it's a little odd to also have it on the "solve" tab of the expanded toolbar (unlike the other icons, I am not able to click and drag that icon to a different section of the toolbar).
- Tooltips do not appear when hovering over or selecting icons within the extended toolbar.
- The "maximize" button on the minimized specs menu (located in the upper-right of that visual) does not seem selectable. Additionally, this minimized window does not overlap in a readable way with the game elements on the left-hand side of the screen.

pic5

**Comment by sbutkovi** [comment]
- When the fancy graphics settings are on, clicking on many of the buttons in the extended toolbar resets the background bubbles.
- Despite the text on the extended toolbar implying otherwise, tapping an icon within the extended toolbar to use it does not always work. Sometimes it does - I'm still working on figuring out when it works correctly (is it when the icon is already on the main toolbar? when the icon has been on the main toolbar within this "reset" session? something else?). However, for a lot of the time, I have to click an icon within the extended toolbar twice in order to use it.
- For icons on the main toolbar and the extended toolbar, I noticed that (unlike the current toolbar) the background color does not shift when the icon is hovered over, and the icon color does not change when the icon is clicked. Is this new behavior the intended behavior?
- Added icons do correctly have text labels when Quick Help is turned on - it could be useful to have these labels also appear when the toolbar is extended (if it can be done elegantly).

pic1

- The annotations menu has issues with updating itself (for example, after adding layers); it only updates what layers are shown when the annotations menu is closed and then opened again. Notable that this buggy behavior is not present on the live website. Additionally, it's slightly odd that the menu appears on the left side instead of the right side like on the live website (a slightly more problematic aspect is that part of the window is falling off the bottom of the screen); it does not seem like I am able to adjust where this menu appears.

pic2

- As I was casually clicking on things, I managed to do something weird - not sure exactly what caused it . . . I might have been trying to close the annotation menu after browsing the extended toolbar for a bit? Note how the annotation icon has moved, some of the icons are falling off of the toolbar, and that some of the toolbar icons were removed (for reference - I was not trying to remove or add any icons when the error occurred, so there _should_ be the same icons here as in my previous screenshot - but obviously there are not).

pic4

pic5

**Comment by sbutkovi** [comment]
- I'm not fond of having my toolbar icons be "kicked" automatically when I change from having large browser window to a small window, especially since I have to manually replace all of them by hand. It also seems arbitrary to me to force players to use a "symmetric" distribution on their toolbars. For example, I can't place 12 icons to the right of the palette - instead, I have to put 6 on the left and 6 on the right.
- I kind of miss having the copy and paste functionalities being located within a single menu instead of as two separate buttons (as implemented here). Yes, I realize that I could go into the extended menu and use copy/paste there, but doing so is a total of ~4 button clicks instead of ~2.
- The PiP icon is available within switch puzzles but not non-switch puzzles. The effect of this is that (1) I can't add PiP to my main toolbar when in a non-switch puzzle, and (2) if I add PiP to my toolbar in a switch puzzle and then go to a non-switch puzzle, the PiP icon doesn't show up (and instead shows a blank space) - see below snapshot (I had added PiP on the left)

pic1
 
- The game freezes when trying to enter the puzzle maker (from the main Dev website - I haven't tried beam to puzzle maker yet):

pic2

**Comment by luxaritas** [comment]
> However, with a lot of poking and prodding (alternating between trying to stretch the various stretchable corners), it is possible to make the window smaller, but only by a few pixels each time (see below). There is similar behavior when trying to increase the size of the window - it can only be stretched larger by a few pixels at a time and works most consistently when alternating between corners.

From my experimentation, it looks like it has something to do with drag speed. The 3D window works correctly afaik, so if this is updated to use the same code as that, hopefully that should resolve that issue

> The settings menu does not recenter when the browser window is resized (here, the browser window was made smaller)

We don't necessarily want it to recenter if you intentionally move it to be not in the center. Again, the 3D window behaves correctly, and using that implementation should make this work as intended.

> When the browser window is vertically small, some of the game elements overlap with each other. Notably, the natural/target mode buttons (and everything below them within the game) are not visible, and there is no way for the user to "scroll" and see them (which could potentially be an issue for small screens, such as mobile screens).

The height you have it set to is far shorter than we've ever supported. There is a point where there's really not much we can do. If we could come up with a way to make it actually work, great, but I'm inclined to say the behavior with the new toolbar is good enough for now (at least, I would argue it hasn't gotten significantly worse than it is currently)

> There's something jarring about the text and/or fonts when Quick Help is brought up

Aside from the "g"s being cut off, I don't personally see any issues here (and afaik it hasn't changed at all from the current version). If you have some additional detail on what seems off to you, let us know.

> From my understanding, the icons that are already on the toolbar are supposed to be "greyed out" in the selection menu

I don't think that should be the case - I would argue that we should still support having them be clickable in the expanded toolbar even if they're already in the top toolbar for consistency.

> The color behavior of the palette when the toolbar is extended is inconsistent. Note how the following screenshot has typical palette coloring, but the previous screenshot has the palette greyed out. Along those lines, it seems odd that the boosters icon is no longer greyed out in the extended menu . . .

I think what's happened here is that the disable tools behavior was triggered - namely, by trying to click and drag the toolbar, which on the current version is necessary because if there isn't enough room for all the tools, we put arrows on either side and let you drag the entire toolbar (and we disable it while dragging to prevent tools from being accidentally triggered). It seems like with this update, this behavior is substantially broken. If the intended behavior is that we only allow as many tools as fits on the screen, this behavior should be removed - though I believe there are other scenarios in which the toolbar tools are disabled, in which case we need to update the disabling logic to properly disable all tools (not just the pallette) - and probably disable expanding the toolbar as well.

> Is the boosters button intended to (1) always be on the toolbar, and (2) be in its own uneditable section of the toolbar? If so, it's a little odd to also have it on the "solve" tab of the expanded toolbar (unlike the other icons, I am not able to click and drag that icon to a different section of the toolbar).

This is something I mentioned in the code review itself - it shouldn't have its own bay.

> The "maximize" button on the minimized specs menu (located in the upper-right of that visual) does not seem selectable. Additionally, this minimized window does not overlap in a readable way with the game elements on the left-hand side of the screen.

Now that the window is floating and resizable, I think we should remove the "minimized view" entirely. Instead, we should have improved layout for the main window. Eg, once the height is greater than the width, it change to stack vertically instead of horizontally.

> For icons on the main toolbar and the extended toolbar, I noticed that (unlike the current toolbar) the background color does not shift when the icon is hovered over, and the icon color does not change when the icon is clicked. Is this new behavior the intended behavior?

We haven't explicitly discussed it, but I think the desired behavior here should be that the icon brightens (when hovered, and darkens when pressed

> The annotations menu has issues with updating itself (for example, after adding layers); it only updates what layers are shown when the annotations menu is closed and then opened again. Notable that this buggy behavior is not present on the live website. Additionally, it's slightly odd that the menu appears on the left side instead of the right side like on the live website (a slightly more problematic aspect is that part of the window is falling off the bottom of the screen); it does not seem like I am able to adjust where this menu appears.

This dialog should be changed to be a floating window

> I'm not fond of having my toolbar icons be "kicked" automatically when I change from having large browser window to a small window, especially since I have to manually replace all of them by hand

I guess we need to decide what we want the behavior here to be. I see two options:
1) What I had mentioned above, where we only allow as many tools as fit on the screen. If you make your window shorter, we hide tools on the edges. However, if you make the window larger, the hidden tools should show back up
2) Use behavior like the current site. If you have too many to fit on the screen, add arrows and allow the toolbar to be dragged left and right

> I kind of miss having the copy and paste functionalities being located within a single menu instead of as two separate buttons (as implemented here). Yes, I realize that I could go into the extended menu and use copy/paste there, but doing so is a total of ~4 button clicks instead of ~2.

There's not really anything we can do about that. We explicitly want to avoid having more menus - we're running out of screen space, and we need to implement new tools in a way that doesn't use more screen space over time, and is consistent with all other tools. If you really need them that often, put them in your hot bar, but otherwise, it would seem odd to need to use these specific tools that often for it to matter much.

> The PiP icon is available within switch puzzles but not non-switch puzzles. The effect of this is that (1) I can't add PiP to my main toolbar when in a non-switch puzzle, and (2) if I add PiP to my toolbar in a switch puzzle and then go to a non-switch puzzle, the PiP icon doesn't show up (and instead shows a blank space) - see below snapshot (I had added PiP on the left)

This is largely working as intended - tools should not be shown if they're not able to be used. Having the PiP button in a single state puzzle makes no sense. That said - I think the behavior should be that if you do add it to your toolbar in a switch puzzle, that when you go to a single state puzzle it should be hidden, but if you go back to a switch puzzle it should show again (ie, keep it in the settings, but don't show it in the UI). When hidden, there shouldn't be a blank space though, of course.

**Comment by sbutkovi** [comment]
> We don't necessarily want it to recenter if you intentionally move it to be not in the center. Again, the 3D window behaves correctly, and using that implementation should make this work as intended.

Ah - I didn't realize the window was movable because the mouse cursor didn't change into the "move me" icon.

> The height you have it set to is far shorter than we've ever supported. There is a point where there's really not much we can do. If we could come up with a way to make it actually work, great, but I'm inclined to say the behavior with the new toolbar is good enough for now (at least, I would argue it hasn't gotten significantly worse than it is currently)

That's fair - I had misremembered what the typical mobile dimensions were, and they seem to all be acceptable (sample screenshot below).

pic1

> Aside from the "g"s being cut off, I don't personally see any issues here (and afaik it hasn't changed at all from the current version). If you have some additional detail on what seems off to you, let us know.

Maybe it's just that it's presented differently (screenshot of the current implementation below). The new one has (1) smaller font size and (2) different capitalization.

pic2

> I don't think that should be the case - I would argue that we should still support having them be clickable in the expanded toolbar even if they're already in the top toolbar for consistency.

Yeah, they should be clickable still. I just thought there was a prior conversation where they were supposed to look grey so you knew at a glance which ones were already on the toolbar. I could be misremembering.

New observation - in puzzles/labs involving codons, the chart interferes with the new setup on the left-hand side. It does not seem that the chart can be moved.

pic3

**Comment by luxaritas** [comment]
> New observation - in puzzles/labs involving codons, the chart interferes with the new setup on the left-hand side. It does not seem that the chart can be moved.
If you hit hide chart, I believe it becomes draggable. Regardless, this is functionality that we don't really control (it's a separate Eternascript), we're not actively using it, and at some point we need to just revamp that entirely (with the codon cycler tool design, etc)

**Comment by sharife** [comment]
A few notes:

- The tool icons appear low-res (blurry). Can you use hi-res pngs or svg instead?
- The width of the expanded toolbar is too large. Can you reduce the max width to 780px?
- The tab labels should use bold text.
- The expanded shelf should animate up/down

**Comment by luxaritas** [comment]
> The width of the expanded toolbar is too large. Can you reduce the max width to 780px?

This made me realize something else: With the current implementation, the width of the hotbar is capped to the width of the expanded toolbar. You should be able to continue adding tools until you run out of horizontal room on the screen, while still keeping the expanded toolbar less wide. This might mean that the hotbar needs to be "outside" the expanded toolbar (ie, not included in the background).

> The expanded shelf should animate up/down

A quick implementation note about this: Keep in mind we have a few different things affecting the location of the toolbar:
1) Screen resizes
2) Toolbar expansion
3) Toolbar autohide (settings -> autohide toolbar - this works similarly to having the task bar autohide in Windows)

We need to make sure that all three are able to be in play at the same time, triggered in no particular order (and potentially even triggered while one of the others is occurring, though that's probably less likely). This is also an area where it might take a little bit of effort to try and consolidate the logic to make sure we're not handling this logic in multiple places. Maybe have one function that's responsible for changing the toolbar location and recalculates the position based on whether the toolbar is expanded or not and whether the toolbar is supposed to be hidden via autohide or not, and then in updateLayout when getting triggered by a resize, we call that function with an extra parameter telling it to "snap" rather than animate to the new location?

**Comment by sbutkovi** [comment]
Not sure what the desired behavior is, but currently when the hotbar is "stuffed" some of the Quick Help labels can fall off the edge (pointing to the offscreen icons):

pic1

pic2

**Comment by luxaritas** [comment]
I think the behavior there should be to not show the tooltips for icons that aren't shown.

File: src/eterna/mode/DesignBrowser/ViewSolutionOverlay.ts [hunk 5]
```diff
@@ -242,7 +242,8 @@ export default class ViewSolutionOverlay extends ContainerObject {
                     filter1_arg2: this._props.solution.nodeID
                     /* eslint-enable camelcase */
                 });
-                this.modeStack.pushMode(new CopyTextDialogMode(solutionURL, 'Solution URL'));
+                // this.modeStack.pushMode(new CopyTextDialogMode(solutionURL, 'Solution URL'));
+                (this.mode as GameMode).showDialog(new CopyTextDialog(solutionURL, 'Solution URL'));
```
[the remark this answers is not in the document]
Comment (flaaudwls) [reply src/eterna/mode/DesignBrowser/ViewSolutionOverlay.ts:246]: fixed

**Comment by sbutkovi** [comment]
Highlighting does not seem to update itself when the toolbar is opened/closed. For example, here the palette was highlighted when the toolbar was closed, and the highlighting stays at the bottom when the toolbar is expanded:

pic1

Analogously, if the toolbar starts expanded when the highlighting is triggered, collapsing the toolbar leaves the highlighting at the top:

pic2

This puzzle can be used to test this behavior: https://eternadev.org/puzzles/10382679

**Comment by sbutkovi** [comment]
When a puzzle is finished, only the palette is greyed out:

pic1

However, for the previous toolbar, all of the toolbar greyed out when the puzzle was completed:

pic2

**Comment by sbutkovi** [comment]
Following up on a previous note: it's worth noting that even though the palette highlighting doesn't "travel" when the toolbar is opened/closed, the arrows to the palette do - therefore, there might be a way of making the highlighting update correspondingly?

pic1

**Comment by sbutkovi** [comment]
Observation - for [Nucleotide Mixer 1](https://eternadev.org/puzzles/6502927) the toolbar shows at the bottom (see screenshot), whereas in the original implementation the toolbar was not visible. I'm assuming this new behavior is fine, but I still wanted to record this observation here.

pic1

File: src/eterna/mode/DesignBrowser/ViewSolutionOverlay.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (pragyavaishanav) [reply src/eterna/mode/DesignBrowser/ViewSolutionOverlay.ts:246]: fixed

File: src/eterna/mode/DesignBrowser/ViewSolutionOverlay.ts [hunk 6]
```diff
@@ -268,7 +269,8 @@ export default class ViewSolutionOverlay extends ContainerObject {
                     filter1_arg1: this._props.solution.playerName
                     /* eslint-enable camelcase */
                 });
-                this.modeStack.pushMode(new CopyTextDialogMode(playerURL, 'Player URL'));
+                // this.modeStack.pushMode(new CopyTextDialogMode(playerURL, 'Player URL'));
+                (this.mode as GameMode).showDialog(new CopyTextDialog(playerURL, 'Player URL'));
```
[the remark this answers is not in the document]
Comment (pragyavaishanav) [reply src/eterna/mode/DesignBrowser/ViewSolutionOverlay.ts:273]: done

File: src/eterna/mode/CopyTextDialogMode.ts [hunk 7]
```diff
@@ -44,6 +44,7 @@ export default class CopyTextDialogMode extends AppMode {
 
         inputPanel.pointerDown.connect((e) => e.stopPropagation());
         const target = new DisplayObjectPointerTarget(bg);
+        bg.interactive = false; // kkk
```
[the remark this answers is not in the document]
Comment (pragyavaishanav) [reply src/eterna/mode/CopyTextDialogMode.ts:47]: Yup

File: src/eterna/mode/FeedbackViewMode.ts [hunk 8]
```diff
@@ -306,8 +312,12 @@ export default class FeedbackViewMode extends GameMode {
         );
     }
 
-    private showViewOptionsDialog(): void {
-        this.showDialog(new EternaViewOptionsDialog(EternaViewOptionsMode.LAB));
+    // private showViewOptionsDialog(): void {
+    //     this.showDialog(new EternaViewOptionsDialog(EternaViewOptionsMode.LAB));
+    // }
```
[the remark this answers is not in the document]
Comment (pragyavaishanav) [reply src/eterna/mode/FeedbackViewMode.ts:317]: done

File: src/eterna/mode/PoseEdit/MissionIntroMode.ts [hunk 9]
```diff
@@ -40,6 +40,7 @@ export default class MissionIntroMode extends AppMode {
         new DisplayObjectPointerTarget(background)
             .pointerDown.filter(InputUtil.IsLeftMouse)
             .connect(() => this.play());
+        background.interactive = false; // kkk
```
[the remark this answers is not in the document]
Comment (pragyavaishanav) [reply src/eterna/mode/PoseEdit/MissionIntroMode.ts:43]: done

File: src/eterna/ui/Toolbar.ts [hunk 10]
```diff
@@ -145,676 +386,3015 @@ export default class Toolbar extends ContainerObject {
...
+
+        // This can be used in both puzzlemaker and lab, so we'll create the base button ahead of time
+        // TODO: Maybe these should really be two separate buttons set on different properties?
+        this.makeLayout();
+
+        /*
+        The lower toolbar structure is a HLayoutContainer wrapped in ScrollContainer wrapped in another HLayoutContainer
+        The arrows are in the outer HLayoutContainer, along with the ScrollContainer
+        The actual toolbar content is in the innermost HLayoutContainer
+        */
```
[the remark this answers is not in the document]
Comment (pragyavaishanav) [reply src/eterna/ui/Toolbar.ts:2488]: removed

File: src/eterna/ui/Toolbar.ts (hunk 1 above)
Comment (pragyavaishanav) [reply src/eterna/ui/Toolbar.ts:2512]: created makeCategory

File: src/eterna/ui/Toolbar.ts (hunk 2 above)
Comment (pragyavaishanav) [reply src/eterna/ui/Toolbar.ts:3156]: Yup

File: src/eterna/ui/Toolbar.ts (hunk 3 above)
Comment (pragyavaishanav) [reply src/eterna/ui/Toolbar.ts:3334]: fixed

File: src/eterna/ui/Toolbar.ts (hunk 4 above)
Comment (pragyavaishanav) [reply src/eterna/ui/Toolbar.ts:1783]: fixed

**Comment by sbutkovi** [comment]
Just noticed that the "mark bases" icon looks a little funny (or at least different than before) - the "mark" isn't centered around the circle representing the base. Looks like it's the same for the "base shift" button and some others (find nucleotide, view range). Is this intentional?

pic1

**Comment by sbutkovi** [comment]
- Nucleotide range window's title seems to fall off the edge for the default window size; behavior does not seem affected when I try opening the window within different sizes of browser window.

pic1

- Formatting is funny on the reset window; unlike other pop-up windows, it looks like 2 nested boxes:

pic2

- Annotations seem broken. Annotation tool does not appear to allow for selecting of bases or labeling. However, when I open the annotations summary panel, it appears my clicking/actions were doing something (because there are new annotations shown). However, I am unable to interact with them in any way (showing/hiding, etc.).

pic3

**Comment by sbutkovi** [comment]
I think the HKWS and SVG buttons aren't working for me? I can manipulate them around the toolbar, but clicking them (even multiple times) does not appear to do anything.

**Comment by sbutkovi** [comment]
While I'm editing the icons on the hotbar, there appear to be issues with the icon following my cursor. Specifically, when I click and hold the left-mouse button on one of the icons (not yet on the hotbar) on the bottom of the extended toolbar and then start to move the cursor around, the icon more or less follows if the cursor is within the "dark blue" region just fine, but when I move my cursor outside of that region, the icon mostly does not follow (although occasionally it shows back up). I can still place the icon on the hotbar, so it seems to be just the visual aspect that is affected.

**Comment by sbutkovi** [comment]
I just chatted with DigitalEmbrace about the toolbar, and here are some additional observations:
- It could potentially be nice to expand the "window" where an icon can be inserted instead of replacing another icon, as the current window is a bit small and can be easily missed.
- It appears to be easier to insert icons in the middle of each hotbar section as opposed to the edges. Additionally, it might be easier to do so on one of the sections (for me, it seems easier to edit the right side in this way than the left side).
- While browsing the expanded toolbar with sound on, sometimes there's a clicking sound that happens when you hover over an icon even without clicking it. For DigitalEmbrace, the clicking sound appeared to happen while hovering over the buttons in the bottom of the toolbar that corresponded to ones already on the hotbar. For me, there was no clear pattern in which buttons were making clicking sounds while hovering over them.
- Sometimes the interface lets you insert an icon on one side but not the other, even when there is an equal number of icons currently on both sides. More concretely, we had 4 icons on the left and 4 on the right, and it would only let us insert an icon on the left (only letting us replace icons on the right).

**Comment by sbutkovi** [comment]
With this new toolbar, I personally think that the "freeze" button should always be available in the toolbar and should not be an option within the options menu.

**Comment by sbutkovi** [comment]
There is unusual behavior with the MS2 stamper tool. Notice that (1) the background color for the MS2 stamper icon is differnt than the other icons, and (2) the stamper icon overlaps with another tool's icon:

pic1

This overlapping behavior also occurs when the freeze button is not present:

pic3

The MS2 stamper tool is also not easily draggable to the hotbar. Here, it has gotten stuck (this screenshot is with no mouse buttons being held down).

pic2

**Comment by sbutkovi** [comment]
Just now, I was able to get the switch-specific PiP button to show up in the hotbar of a 1-state puzzle (by putting the PiP button on the hotbar within a switch puzzle and then refreshing the page of the 1-state puzzle):

pic1

**Comment by sbutkovi** [comment]
It's possible to get the toolbar stuck below the bottom of your screen. I can't exactly remember the steps I did to get here the first time I did it, but I remember that I was testing the autohide toolbar option and had turned it on and then off (with some clicking and toolbar actions in between). This particular view happened when the autohide toolbar option was off.

pic1

Refreshing the page restored my toolbar, but it shouldn't get stuck in the first place.

Here's at least one walkthrough that seems to replicate this issue:
- Start with "Autohide toolbar" turned off
- Turn on "Autohide toolbar"
- Expand toolbar
- While toolbar is expanded, turn off "Autohide toolbar" option (when this is turned off, then you should see the toolbar stay expanded)
- Click the downward arrow to collapse the toolbar

Following this order of actions should make your toolbar stuck below the screen (in the current version) - which is not ideal.

**Comment by tkaragianes** [comment]
Slava, I also ran into this bug, but it was tricky to figure out the exact order of operations that would lead to it. I replicated just now with your steps, and it does trigger the bug. Thanks for the legwork figuring out a reliable repro case.

**Comment by sbutkovi** [comment]
DigitalEmbrace ran across a fatal error when using annotations with this new toolbar - I'm reporting the error and screenshot she saw here on her behalf.

Her description: "Testing annotations in puzzlemaker using new toolbar on Dev. Got this message when I clicked on the "3' end" annotation right after I had edited the "5' end" annotation."

Screen Shot 2022-08-02 at 7 30 43 AM

**Comment by sbutkovi** [comment]
There's something odd happening in this latest iteration of the toolbar. I haven't yet determined to what extent these two issues are connected or separate:

1. There's a blank space on the left part of the left-hand hotbar. I think this is due to the buttons of the left-hand hotbar being shifted one "slot" to the right rather than an invisible button, but I'm not entirely sure.

2. The left-hand side of the hotbar does not stay "solid" when the toolbar is expanded and then collapsed again; specifically, the right-hand button of the left-hand hotbar loses its background color. See the following 3 screenshots:

pic2

pic3

pic4

Refreshing or hard refreshing the page does not fix the issue. This issue is also visible on multiple puzzles (for example, https://eternadev.org/game/puzzle/10382363/ and https://eternadev.org/game/puzzle/10191854/).

**Comment by sbutkovi** [comment]
"There's something odd happening in this latest iteration of the toolbar. I haven't yet determined to what extent these two issues are connected or separate:"

Actually, I'm not seeing either of these issues anymore, and the toolbar works really smoothly, so . . . maybe everything is okay? Maybe it was just a fluke.

**Pull request merged**